src/ec2/model.ts

```diff
@@ -46,26 +50,30 @@ export class Ec2ConnectionManager {
         return new DefaultIamClient(this.regionCode)
     }
 
-    protected async getAttachedPolicies(instanceId: string): Promise<IAM.AttachedPolicy[]> {
-        const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
-        if (!IamRole?.Arn) {
-            return []
+    public async getAttachedIamRole(instanceId: string): Promise<IAM.Role> {
+        const IamInstanceProfile = await this.ec2Client.getAttachedIamInstanceProfile(instanceId)
+        if (IamInstanceProfile && IamInstanceProfile.Arn) {
+            const IamRole = await this.iamClient.getIAMRoleFromInstanceProfile(IamInstanceProfile.Arn)
+            return IamRole
         }
+        throw new ToolkitError(`No IAM instance profile attached to instance ${instanceId}`, {
+            code: 'NoIamInstanceProfile',
```

Comment:
If `getAttachedIamRole` returns `undefined`, the caller knows unambiguously that nothing was found. Don't need to throw here. 

A caller may choose to throw, however.

Reply:
Yea, the refactored flow makes use of this exactly.